src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs

```diff
@@ -70,15 +70,17 @@ internal override bool OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializer
             Debug.Assert(options == jsonTypeInfo.Options);
 
             if (!state.SupportContinuation &&
+                jsonTypeInfo.HasSerializeHandler &&
```

Comment:
Not directly related to the fix, but this applies a performance optimization similar to the root optimization methods. `is JsonTypeInfo<T>` checks are expensive so we prefer to guard them with checks that the handler does indeed exist.

Reply:
seems related enough to me since you're reviewing every usage